debian/cloud-init.postinst

```diff
@@ -276,6 +276,51 @@ fix_azure_upgrade_1611074() {
     rm "$tmpf" || :
 }
 
+fix_lp1889555() {
+    local oldver="$1" last_bad_ver="20.2-94-g3d06abc2-0ubuntu1"
+    dpkg --compare-versions "$oldver" le-nl "$last_bad_ver" || return 0
+
+    # if cloud-init's grub module did not run, then it did not break anything.
+    [ -f /var/lib/cloud/instance/sem/config_grub_dpkg ] || return 0
+
+    # Don't do anything unless we have grub
+    [ -x /usr/sbin/grub-install ] || return 0
+
+    # Make sure that we are not chrooted.
+    [ "$(stat -c %d:%i /)" != "$(stat -c %d:%i /proc/1/root/.)" ] && return 0
+
+    # Check if we are in a container, i.e. LXC
+    if systemd-detect-virt --quiet --container || lxc-is-container 2>/dev/null; then
+        return 0
+    fi
+
+    # This bug only applies to NVMe devices
+    [ -e /dev/nvme0 ] || return 0
+
+    db_get grub-pc/install_devices && grub_cfg_dev=${RET} || return 0
+
+    # If the current setting is not the (incorrect) default we expect, this
+    # implies user intervention so leave things alone
+    # XXX: Do we thinkg this is the correct behaviour?
+    [ "$grub_cfg_dev" = "/dev/sda" ] || return 0
```

Comment:
For the sake of risk mitigation to a wider audience, I think that this is an excellent check.

The problem is, if a user has experienced the issue, such as those in this webpage:
https://www.redradishtech.com/display/~jturner/2020/07/30/symbol+'grub_calloc'+not+found+--+how+to+fix+on+AWS
Users were encouraged to select every box in the interactive dpkg prompt to get their system to boot. While it will ensure that their systems will boot, I believe that the debconf database sets grub-pc/install_devices to the final selected entry, regardless if it was the correct option or not.

In the above link, the user notes that theirs has been set to:
/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol01678967bebb56f55-part1

This is wrong. This is selecting a partition of the disk instead of the disk itself, and it might not even be the correct disk.

In the refactor, I took special care to always omit devices with -part[x] in their name. This was resolved nicely with "udevadm info" on the boot disk, since it always returns /dev/disk/by-id values for the disk itself, and never the partitions. 

If you keep this check, you will not fix users who attempted to fix their own machines and did it incorrectly.

Another edge case is say I had two disks. /dev/vda is a scratch disk, and /dev/vdb is my boot disk. /dev/vda was selected, since its a hard coded directory in previous cc_grub_dpkg. But I have now hot removed /dev/vda and not rebooted since, so only /dev/vdb remains. We will still install to /dev/vda, and I'll have a broken system on reboot.

In the above two cases, it would make sense to clear out grub-pc/install_devices and simply run cc_grub_dpkg to determine the actual boot disk, and go from there.

Its tough. This check limits risk to larger audiences, but clearing out the value entirely and re-running cc_grub_dpkg would result in more machines getting fixed. The risk to larger audiences is if there is a regression in the refactored implementation of cc_grub_dpkg that we haven't thought of, we might inadvertently break users who weren't initially broken.

Reply:
> If you keep this check, you will not fix users who attempted to fix their own machines and did it incorrectly.

I consider fixing such machines a non-goal: I don't believe that we can reliably distinguish between "user failed to fix this problem" and "user is intentionally doing something special with grub"; if users have manually intervened in this configuration then, in my opinion, we should be leaving it alone.

> Another edge case is say I had two disks. /dev/vda is a scratch disk, and /dev/vdb is my boot disk. /dev/vda was selected, since its a hard coded directory in previous cc_grub_dpkg. But I have now hot removed /dev/vda and not rebooted since, so only /dev/vdb remains. We will still install to /dev/vda, and I'll have a broken system on reboot.

Similarly, I don't believe we can programmatically distinguish (without coding for this specific case) between an absent scratch disk and an absent disk that was mounted at `/boot` and will be mounted back there once it returns.

> The risk to larger audiences is if there is a regression in the refactored implementation of cc_grub_dpkg that we haven't thought of, we might inadvertently break users who weren't initially broken.

I disagree, the risk is not just that `cc_grub_dpkg` has a bug or regression: `cc_grub_dpkg` is intended to be run once at first boot, before users will have been able to apply much/any of their own configuration. Crucially, because it happens during boot normally, we can be pretty confident that the system is in a bootable state. However, we are now running it at _upgrade_ time, which could occur any time during an instance's lifecycle: the system's state is a lot less predictable, and so could invalidate the preconditions on which `cc_grub_dpkg` relies (e.g. `/boot` is mounted); this wouldn't be a bug in `cc_grub_dpkg`, it would be a bug in how we decided to call it (i.e. in this script).

This is particularly true if we end up publishing this fix, or something like it, to `-security`: unattended-upgrades will install it without user intervention, so we cannot rely on them doing anything to put the system into a "good" state before applying upgrades.